lib/internal/url.js

```diff
@@ -839,20 +826,18 @@ class URL {
     this.#context.hash_start = hash_start;
     this.#context.scheme_type = scheme_type;
 
+    // If URLSearchParams is in use, store the search there, not here.
     if (this.#searchParams) {
       if (this.#context.hasSearch) {
-        setURLSearchParams(this.#searchParams, this.search);
-      } else {
-        setURLSearchParams(this.#searchParams, undefined);
+        setURLSearchParams(this.#searchParams, this.#getContextSearch());
+        this.#updateContext(bindingUrl.update(href, updateActions.kSearch, ''));
       }
     }
   }
 
-  toString() {
-    return this.#context.href;
-  }
-
   get href() {
+    // URLSearchParams is used on demand, and stores the search if it's used.
+    if (this.#searchParams) return bindingUrl.update(this.#context.href, updateActions.kSearch, this.#searchParams.toString());
```

Comment:
This makes an extra C++ call which significantly impacts the performance. Since almost all getters relies on href attribute, this impacts all getters.

Reply:
All the getters use the internal `#context.href` I believe, this should only impact `href`, `toString` and `toJSON`?

That being said, is there a better way to compose the URL that is safe?

Reply:
Is it considered safe enough to just use `StringPrototypeSlice` and drop `this.#searchParams.toString()` in before `this.#context.hash_start`?

Reply:
It might be I'm not sure. Node has a WPT test runner which runs with the test suite.